[GLT-3314] Precise Case Report

avarsava opened the pull request:
Missing some features:
- Is there a way to get Samples by Tissue Origin, parent etc without using String.contains() etc?
- How does one get a Sample's Group ID?

avarsava committed:
Untested, missing features

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+     */
+    public Predicate<SampleDto> predicate() {
+      return sample -> {
+        if (SAMPLE_CATEGORY_IDENTITY.equals(sample.getSampleType())) return false;
```
While this will work since the class name is the same, it's ideal to compare to the "Sample Category" attribute. Can use `getAttribute(ATTR_CATEGORY, sample)` from `SampleUtils` for this

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+    table = new LinkedList<>();
+
+    // Add headers
+    table.add(
+        Lists.newArrayList(
+            "ID",
+            "Screening/Randomization Serum",
```
Since the columns never change, they should be defined statically and returned from `getColumns()`. See `ProjectSequencingReport` for a simple example

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+    allPreciseIdentities =
+        allPreciseSamples
+            .stream()
+            .filter(s -> s.getSampleType().equals(SAMPLE_CATEGORY_IDENTITY))
```
Similarly to above, ideal to compare to Sample Category attribute

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+                .stream()
+                .filter(
+                    s ->
+                        s.getName().startsWith(identity.getName())) // Get children of this Identity
```
Never parse the name for information if it can be avoided. You can use `SampleUtils#getParent(String, String, Map<String, SampleDto)` to find the sample's Identity. `SampleUtils#mapSamplesById` is also useful to create the necessary map.

avarsava replied:
Hi Dillan, in this line I have an identity (that's identity) and need to get all of its children. It doesn't look like there's a 'getChildren' to mirror getParent, is there another way to achieve this or should I implement a 'getChildren'?

djcooke replied:
There is no `getChildren` method. In the report I'm working on, I'm getting all of the descendants for each identity using this method:

```
  private Map<String, List<SampleDto>> mapChildrenByIdentityId() {
    Map<String, List<SampleDto>> childrenByIdentityId = new HashMap<>();
    for (SampleDto sample : samplesById.values()) {
      if (!SAMPLE_CATEGORY_IDENTITY.equals(getAttribute(ATTR_CATEGORY, sample))) {
        SampleDto identity = getParent(sample, SAMPLE_CATEGORY_IDENTITY, samplesById);
        if (!childrenByIdentityId.containsKey(identity.getId())) {
          childrenByIdentityId.put(identity.getId(), new ArrayList<>());
        }
        childrenByIdentityId.get(identity.getId()).add(sample);
      }
    }
    return childrenByIdentityId;
  }
```

(and then I have another list filtered to include only the identities)

You could do something similar, or go the other way, using `SampleDto#getChildren` and working your way down the tree. Going up is usually simpler though due to branches and variable hierarchy

avarsava replied:
Thanks! Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+        for (String origin : tissueOrigins) {
+          row.add(
+              String.valueOf(
+                  samplesAtCurrentTime.stream().filter(s -> s.getName().contains(origin)).count()));
```
Use `SampleUtils#getAttribute` and/or `SampleUtils#getUpstreamAttribute` to get tissue origin instead of parsing name

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+                  .filter(s -> s.getName().startsWith(identity.getName()))
+                  .filter(s -> s.getSampleType().equals(SAMPLE_CLASS_SLIDE))
+                  .filter(TimePoint.BX.predicate())
+                  .filter(s -> true) // TODO: How to get Group ID? Need that for POS/NEG
```
`SampleUtils#getAttribute` and/or `SampleUtils#getUpstreamAttribute`

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+          String.valueOf(
+              allPreciseSamples
+                  .stream()
+                  .filter(s -> s.getName().startsWith(identity.getName()))
```
Use `SampleUtils#getParent`

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+public class PreciseIdentityReport extends TableReport {
+  public static final String REPORT_NAME = "preciseidentity";
+  public static final String CATEGORY = REPORT_CATEGORY_COUNTS;
+  private String start;
+  private String end;
```
These don't seem to be used for anything except report title, which should use the current date instead

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java:
```diff
@@ -0,0 +1,233 @@
+import org.apache.commons.cli.ParseException;
+
+public class PreciseIdentityReport extends TableReport {
+  public static final String REPORT_NAME = "preciseidentity";
```
Should be added to report option list in `Main#getMainOptions`

avarsava replied:
Updated

avarsava committed:
Rename to PRECISE Case Report'

avarsava committed:
Remove options, use localdate

avarsava committed:
Test against ATTR_CATEGORY

avarsava committed:
and again

avarsava committed:
Fix columns

avarsava committed:
Fix origin comparison

avarsava committed:
lookup group ID

avarsava committed:
Get children in a more reliable manner

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,229 @@
+              String.valueOf(
+                  samplesAtCurrentTime
+                      .stream()
+                      .filter(s -> origin.equals(getAttribute(ATTR_TISSUE_ORIGIN, s)))
```
Note that tissue origin is only present on tissue samples, so as is, this will only count tissues. I'm not sure if that's what you want, but if you do want to include other classes downstream from tissues, you can also check `SampleUtils#getUpstreamAttribute`

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,229 @@
+                  .stream()
+                  .filter(s -> s.getSampleType().equals(SAMPLE_CLASS_SLIDE))
+                  .filter(TimePoint.BX.predicate())
+                  .filter(s -> getAttribute(ATTR_GROUP_ID, s).equals("POS"))
```
Group ID will only be present on the sample it's set on, but it should affect children as well, so parents should be checked using `SampleUtils#getUpstreamAttribute` too

avarsava replied:
Updated

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,229 @@
+      row.add(
+          String.valueOf(
+              allPreciseSamplesByIdentity
+                  .get(identity) // Get children of this identity
```
Map key is identity ID rather than the identity itself. Same other places this is used

avarsava replied:
Updated

avarsava committed:
Filter fixes

avarsava committed:
Yoda tests

avarsava committed:
Empty list fix, broken String Array

avarsava committed:
Ilinca says no tissue descendants

avarsava committed:
Fix row count issue

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    /**
+     * Assumes that PRECISE samples will always have _nn_ in the name, and that Tissue Origin will
+     * never be "nn". Should hopefully be a safe assumption since this is a biobanking project where
+     * tracking tissue origin is critical.
```
I don't think anything depends on this assumption anymore?

avarsava replied:
Fixed!

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,259 @@
+          }
+        }
+      }
+      if (potentialSamplesAtCurrentTime != null) {
```
I think this means that it only looks for time points 6 and 7 if there are samples for time point 5? Might want to handle this differently in-case time point 5 gets skipped for some reason. Then `potentialSamplesAtCurrentTime` could be moved to a narrower scope as well.

avarsava replied:
The result would be the same every loop, but i cleaned up the logic a little to save some wasted time

djcooke replied:
Better not to make assumptions about future samples

djcooke reviewed (commented):
Please add a file to `/docs/` to describe the report similar to the others

avarsava committed:
Fix up logic

avarsava committed:
Add doc

avarsava commented:
> Please add a file to `/docs/` to describe the report similar to the others

Added a link to a wiki page and converted the instructions for this report to a wiki page

djcooke commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,252 @@
+    for (SampleDto identity : allPreciseIdentities) {
+      List<String> row = new LinkedList<>();
+      row.add(identity.getName());
+      potentialSamplesAtCurrentTime = allPreciseSamplesByIdentity.get(identity.getId());
+      if (potentialSamplesAtCurrentTime == null) {
```
`potentialSamplesAtCurrentTime` isn't specific to current time, but it's not really used for anything either. Could check `if (allPreciseIdentitiesByIdentity.hasKey(identity.getId))` instead

avarsava replied:
Fixed

djcooke commented on docs/PreciseCaseReport.md:
```diff
@@ -0,0 +1,2 @@
+# PRECISE Case Report
+See [wiki link](https://wiki.oicr.on.ca/display/GSI/PRECISE+Case+Report) for report documentation.
```
Is there a reason not to include the documentation here like the other reports?

avarsava replied:
PRECISE has its doc link to a wiki page, figured this would be of similar privacy level

djcooke replied:
I'm not sure why that is. Maybe because project-specific information is revealed - types of samples received and labs they're received from. Whether it's on the wiki or in this repo, documentation should be in the same format as other reports though

avarsava replied:
Rewrote the wiki page. Please let me know if there is any further information I should include on the page.

avarsava committed:
use containsKey

avarsava committed:
Fix reversed bool

apmasell commented on src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java:
```diff
@@ -0,0 +1,250 @@
+      row.add(identity.getName());
+      if (!allPreciseSamplesByIdentity.containsKey(identity.getId())) {
+        row.addAll(
+            Lists.newArrayList(
```
It would be a lot easier to read with `Collections.nCopies`

avarsava replied:
Fixed

avarsava committed:
use nCopies
